### ISSUE TITLE
False positive in arc_with_non_send_sync

### ISSUE THREAD

**Issue by popzxc** [rust-lang/rust-clippy#11382]
### Summary

This lint triggers when a non-send type is wrapped in `Arc`, even if this `Arc` is then used by the type that is `Send`.
The reproducer example is messy, but still, probably lint shouldn't trigger there.

[Playground link for convenience.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9a5f478cd89ab7f365995ea964233eba)

### Lint Name

arc_with_non_send_sync

### Reproducer

I tried this code:

```rust
use std::{sync::Arc, marker::PhantomData, cell::Cell};

#[derive(Default)]
struct ResourceNotSend {
    _not_send: PhantomData<Cell<()>>,
}

impl ResourceNotSend {
    fn threadsafe_method(&self) {
        println!("huh");
    }
}

struct Worker {
    resource: Arc<ResourceNotSend>
}

unsafe impl Send for Worker {}
unsafe impl Sync for Worker {}

impl Worker {
    fn foo(&self) {
        self.resource.threadsafe_method();
    }
}


fn main() {
    let resource = Arc::new(ResourceNotSend::default());
    let worker = Worker { resource: resource.clone() };
    std::thread::spawn(move || {
        worker.foo();
    }).join().unwrap();
}
```

I saw this happen:

```
warning: usage of an `Arc` that is not `Send` or `Sync`
  --> src/main.rs:29:20
   |
29 |     let resource = Arc::new(ResourceNotSend::default());
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the trait `Sync` is not implemented for `ResourceNotSend`
   = note: required for `Arc<ResourceNotSend>` to implement `Send` and `Sync`
   = help: consider using an `Rc` instead or wrapping the inner type with a `Mutex`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arc_with_non_send_sync
   = note: `#[warn(clippy::arc_with_non_send_sync)]` on by default

```

I expected to see this happen:

No lint spawned.

### Version

```text
rustc 1.73.0-nightly (08d00b40a 2023-08-09)
binary: rustc
commit-hash: 08d00b40aef2017fe6dba3ff7d6476efa0c10888
commit-date: 2023-08-09
host: aarch64-apple-darwin
release: 1.73.0-nightly
LLVM version: 17.0.0
```

**Comment by filiptibell** [comment]
Yep, looks like this happens when using tokio's `Mutex` too, so lots of these lints just started showing up for us with the `1.72` release :(
